cirq-google/cirq_google/devices/grid_device.py

```diff
@@ -373,6 +461,103 @@ def from_proto(cls, proto: v2.device_pb2.DeviceSpecification) -> 'GridDevice':
 
         return GridDevice(metadata)
 
+    def to_proto(
+        self, out: Optional[v2.device_pb2.DeviceSpecification] = None
+    ) -> v2.device_pb2.DeviceSpecification:
+        """Serializes the GridDevice to a DeviceSpecification.
+
+        Args:
+            out: Optional DeviceSpecification to be populated. Fields are populated in-place.
+
+        Returns:
+            The populated DeviceSpecification if out is specified, or the newly created
+            DeviceSpecification.
+        """
+        qubits = self._metadata.qubit_set
+        unordered_pairs = [tuple(pair_set) for pair_set in self._metadata.qubit_pairs]
+        pairs = [
+            (pair[0], pair[1]) if pair[0] <= pair[1] else (pair[1], pair[0])
+            for pair in unordered_pairs
+        ]
+        gateset = self._metadata.gateset
+        gate_durations = self._metadata.gate_durations
+
+        if out is None:
+            out = v2.device_pb2.DeviceSpecification()
+
+        # If fields are already filled (i.e. as part of the old DeviceSpecification format), leave
+        # them as is. Fields populated in the new format do not conflict with how they were
+        # populated in the old format.
+        # TODO(#5050) remove empty checks below once deprecated fields in DeviceSpecification are
+        # removed.
+
+        if len(out.valid_qubits) == 0:
+            known_devices.populate_qubits_in_device_proto(qubits, out)
+        if len(out.valid_targets) == 0:
+            known_devices.populate_qubit_pairs_in_device_proto(pairs, out)
+        _serialize_gateset_and_gate_durations(
+            out, gateset, {} if gate_durations is None else gate_durations
+        )
+        _validate_device_specification(out)
+
+        return out
+
+    @classmethod
+    def from_device_information(
+        cls,
+        *,
+        qubit_pairs: Collection[Tuple[cirq.GridQubit, cirq.GridQubit]],
+        gateset: cirq.Gateset,
+        gate_durations: Optional[Mapping['cirq.GateFamily', 'cirq.Duration']] = None,
+    ) -> 'GridDevice':
+        """Constructs a GridDevice using the device information provided.
+
+        This is a convenience method for constructing a GridDevice given partial gateset and
+        gate_duration information: for every distinct gate, only one representation needs to be in
+        gateset and gate_duration. The remaining representations will be automatically generated.
+
+        For example, if the input gateset contains only `cirq.PhasedXZGate`, and the input
```

Comment:
Do you have a test for this expansion of compatible gatesets?

Reply:
It's tested in [`test_device_from_device_information_equals_device_from_proto`](https://github.com/quantumlib/Cirq/blob/4837660ed7d9aa2d415bd632c381abd5926d1e85/cirq-google/cirq_google/devices/grid_device_test.py#L473) by checking that the generated device matches the one from `from_proto()`. The gateset in the device from `from_proto()` is verified in [`test_grid_device_from_proto`](https://github.com/quantumlib/Cirq/blob/4837660ed7d9aa2d415bd632c381abd5926d1e85/cirq-google/cirq_google/devices/grid_device_test.py#LL278C46-L278C46).